### PR TITLE
Add recent block gas data to metamask state

### PR DESCRIPTION
**Pull request by danfinlay**
Will allow more intelligent gas price suggestions.

On the `state.metamask.recentGasPrices`, there is now an array of arrays.
The first array of arrays represents recent blocks with transactions.
The arrays contained contain hex-encoded strings, representing
gas prices for recent successfully mined transactions.

These arrays can be digested to create recent successful gas price averages.

### CODE REVIEW

**Comment by Zanibas** [comment]
To clarify, this is a purely state-adding PR, with no UI changes yet?

**Comment by danfinlay** [comment]
Correct, this merely enables the more advanced gas inputs we have plans for.

File: app/scripts/metamask-controller.js [hunk 1]
```diff
@@ -187,25 +195,30 @@ module.exports = class MetamaskController extends EventEmitter {
...
+    this.personalMessageManager.memStore.subscribe(sendUpdate)
+    this.typedMessageManager.memStore.subscribe(sendUpdate)
+    this.keyringController.memStore.subscribe(sendUpdate)
+    this.preferencesController.store.subscribe(sendUpdate)
+    this.recentGasPrices.store.subscribe(sendUpdate)
+    this.addressBookController.store.subscribe(sendUpdate)
+    this.currencyController.store.subscribe(sendUpdate)
+    this.noticeController.memStore.subscribe(sendUpdate)
+    this.shapeshiftController.store.subscribe(sendUpdate)
+    this.infuraController.store.subscribe(sendUpdate)
```
Comment (danfinlay) [inline app/scripts/metamask-controller.js]: @kumavis do you think we could replace all of these above with a simple `this.store.subscribe(sendUpdate)`? As long as these stores are already being used to `updateState()` on `this.store` (which they are above), isn't this block somewhat redundant? Is there something I'm missing?

File: app/scripts/controllers/recent-gas-prices.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      return
+    }
+
+    const gasPrices = newBlock.transactions
+    .map(tx => tx.gasPrice)
+
+    const state = this.store.getState()
+    state.recentGasPrices.push(gasPrices)
+
+    while (state.recentGasPrices.length > this.historyLength) {
```
Comment (Zanibas) [inline app/scripts/controllers/recent-gas-prices.js]: nitpicking here, but can't we use `splice` here?

File: app/scripts/metamask-controller.js (hunk 1 above)
Comment (Zanibas) [reply app/scripts/metamask-controller.js]: This is part of the obs-store refactor that's at the bottom of the PR list at the moment, right?

**Review by 2-am-zzz — APPROVED** [review]
lgtm, non-blocking comments

File: app/scripts/controllers/recent-gas-prices.js (hunk 2 above)
Comment (danfinlay) [reply app/scripts/controllers/recent-gas-prices.js]: Example?

**Comment by danfinlay** [comment]
@chikeichan The tests fail, but I want to confirm that these are the failures you expect on `NewUI-flat`?

**Comment by danfinlay** [comment]
Closing in favor of #2623.

**Pull request closed without merging**